### DISCUSSION TITLE
Dockerfile isn't buildable

### DISCUSSION THREAD

**Discussion by aftonsteps** [ThinkR-open/golem#588] [Q&A]
Hi all. I'm having trouble building a Docker image based upon the Dockerfile create by `golem::add_dockerfile()`. I'm relatively new to Docker, so I might be doing something wrong. I appreciate any help.

I'm trying to create a Dockerfile, and then build locally to make sure it works. Sorry for the lengthy post. I tried a few things, which all resulted in errors.

First, I tried using the Dockerfile as it is produced by `golem:add_dockerfile()`. This produced this Dockerfile:

```
FROM rocker/r-ver:3.6.1
RUN apt-get update && apt-get install -y  git-core libcurl4-openssl-dev libgit2-dev libicu-dev libssl-dev libxml2-dev make pandoc pandoc-citeproc python-minimal zlib1g-dev && rm -rf /var/lib/apt/lists/*
RUN echo "options(repos = c(CRAN = 'https://cran.rstudio.com/'), download.file.method = 'libcurl')" >> /usr/local/lib/R/etc/Rprofile.site
RUN R -e 'install.packages("remotes")'
... <various R packages to install> ...
```

There was a pretty immediate error when trying to build, because `libcurl4-gnutls-dev` is missing: 

```
The following packages have unmet dependencies:

 libgit2-dev : Depends: libcurl4-gnutls-dev but it is not going to be installed

E: Unable to correct problems, you have held broken packages.
```

Then I tried to add `libcurl4-gnutls-dev`, but it looks like it can't coexist with `libcurl4-openssl-dev`:

```
The following packages have unmet dependencies:

 libcurl4-gnutls-dev : Conflicts: libcurl4-openssl-dev but 7.52.1-5+deb9u12 is to be installed

 libcurl4-openssl-dev : Conflicts: libcurl4-gnutls-dev but 7.52.1-5+deb9u12 is to be installed

E: Unable to correct problems, you have held broken packages.
```

Then I tried removing `libcurl4-openssl-dev` entirely, and just leaving `libcurl4-gnutls-dev`. In this case, I was able to get past the immediate error, and at least get to the portion of the build that installs R packages. However, there was a problem with `gert`:

```
clone.c: In function ‘R_git_remote_ls’:

clone.c:436:11: error: too many arguments to function ‘git_remote_connect’

   bail_if(git_remote_connect(remote, GIT_DIRECTION_FETCH, &callbacks, NULL, NULL), "git_remote_connect");

           ^~~~~~~~~~~~~~~~~~

In file included from /usr/include/git2/clone.h:14:0,

                 from /usr/include/git2.h:19,

                 from utils.h:1,

                 from clone.c:6:

/usr/include/git2/remote.h:247:17: note: declared here

 GIT_EXTERN(int) git_remote_connect(git_remote *remote, git_direction direction, const git_remote_callbacks *callbacks, const git_strarray *custom_headers);

                 ^~~~~~~~~~~~~~~~~~

/usr/local/lib/R/etc/Makeconf:167: recipe for target 'clone.o' failed

make[1]: *** [clone.o] Error 1

make[1]: Leaving directory '/tmp/RtmpuXXgMw/R.INSTALL318370272f6/gert/src'

ERROR: compilation failed for package ‘gert’

* removing ‘/usr/local/lib/R/site-library/gert’

cat: usethis.out: No such file or directory

Error: Failed to install 'golem' from GitHub:

  (converted from warning) installation of one or more packages failed,

  probably ‘gert’, ‘usethis’

Execution halted

The command '/bin/sh -c Rscript -e 'remotes::install_github("ThinkR-open/golem@558a715bfd601468cd7131a719014e3a256d9390")'' returned a non-zero code: 1
```

Then, I read [here](https://github.com/r-hub/sysreqsdb/issues/77) about what might (?) be a related issue. I tried using the CRAN PPA instead, with a new Dockerfile:

```
FROM rocker/r-ver:3.6.1
RUN apt-get update -y && apt-get install software-properties-common -y && add-apt-repository ppa:cran/libgit2 && apt-get install -y  git-core libcurl4-gnutls-dev libicu-dev libssl-dev libxml2-dev make pandoc pandoc-citeproc python-minimal zlib1g-dev && rm -rf /var/lib/apt/lists/*
RUN echo "options(repos = c(CRAN = 'https://cran.rstudio.com/'), download.file.method = 'libcurl')" >> /usr/local/lib/R/etc/Rprofile.site
```

This results in a new error for `gert`:

```
* installing *source* package ‘gert’ ...

** package ‘gert’ successfully unpacked and MD5 sums checked

** using staged installation

Using PKG_CFLAGS=

Using PKG_LIBS=-lgit2

----------------------------- ANTICONF -------------------------------

Configuration failed to find libgit2 library. Try installing:

 * brew: libgit2 (MacOS)

 * deb: libgit2-dev (Debian, Ubuntu, etc)

 * rpm: libgit2-devel (Fedora, CentOS, RHEL)

If libgit2 is already installed, check that 'pkg-config' is in your

PATH and PKG_CONFIG_PATH contains a libgit2.pc file. If pkg-config

is unavailable you can set INCLUDE_DIR and LIB_DIR manually via:

R CMD INSTALL --configure-vars='INCLUDE_DIR=... LIB_DIR=...'

-------------------------- [ERROR MESSAGE] ---------------------------

<stdin>:1:18: fatal error: git2.h: No such file or directory

compilation terminated.

----------------------------------------------------------------------

ERROR: configuration failed for package ‘gert’

* removing ‘/usr/local/lib/R/site-library/gert’

Error: Failed to install 'golem' from GitHub:

  (converted from warning) installation of package ‘gert’ had non-zero exit status

Execution halted

The command '/bin/sh -c Rscript -e 'remotes::install_github("Thinkr-open/golem@aaae5c8788802a7b4aef4df23691902a286dd964")'' returned a non-zero code: 1
```

I really have no idea how to make this build. 🙁  Thank you for any help.

**Comment by jacciz** [comment]
I did a workaround when I made my dockerfile. I was having issues with libraries not installing, like libxml2-dev. Instead of using the dockerfile Golem made, I took some of the code and made my own via rocker-versioned2. You'll have 2 extra folders - stacks and scripts that Golem may warn - these are for building images.


`FROM rocker/shiny-verse:4.0.3
#### Rocker image includes R, Shiny, and tidyverse

LABEL org.label-schema.license="GPL-2.0" \
      org.label-schema.vcs-url="https://github.com/rocker-org/rocker-versioned" \
      org.label-schema.vendor="Rocker Project" \
      maintainer="Carl Boettiger <cboettig@ropensci.org>"


RUN /rocker_scripts/install_geospatial.sh

COPY . ./app

RUN echo "options(repos = c(CRAN = 'https://cran.rstudio.com/'), download.file.method = 'libcurl')" >> /usr/local/lib/R/etc/Rprofile.site
RUN R -e 'install.packages("remotes")'
RUN R -e 'remotes::install_github("r-lib/remotes", ref = "97bbf81")'
RUN Rscript -e 'remotes::install_version("glue",upgrade="never", version = "1.4.2")'
RUN Rscript -e 'remotes::install_version("stringr",upgrade="never", version = "1.4.0")'
RUN Rscript -e 'remotes::install_version("processx",upgrade="never", version = "3.4.5")'
RUN Rscript -e 'remotes::install_version("htmltools",upgrade="never", version = "0.5.1.1")'
RUN Rscript -e 'remotes::install_version("pkgload",upgrade="never", version = "1.2.0")'
RUN Rscript -e 'remotes::install_version("testthat",upgrade="never", version = "3.0.2")'
RUN Rscript -e 'remotes::install_version("DBI",upgrade="never", version = "1.1.1")'
RUN Rscript -e 'remotes::install_version("htmlwidgets",upgrade="never", version = "1.5.3")'
RUN Rscript -e 'remotes::install_version("knitr",upgrade="never", version = "1.31")'
RUN Rscript -e 'remotes::install_version("attempt",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("shiny",upgrade="never", version = "1.6.0")'
RUN Rscript -e 'remotes::install_version("data.table",upgrade="never", version = "1.14.0")'
RUN Rscript -e 'remotes::install_version("sf",upgrade="never", version = "0.9-7")'
RUN Rscript -e 'remotes::install_version("leaflet",upgrade="never", version = "2.0.4.1")'
RUN Rscript -e 'remotes::install_version("config",upgrade="never", version = "0.3.1")'
RUN Rscript -e 'remotes::install_version("rmarkdown",upgrade="never", version = "2.7")'
RUN Rscript -e 'remotes::install_version("pool",upgrade="never", version = "0.1.6")'
RUN Rscript -e 'remotes::install_version("shinyWidgets",upgrade="never", version = "0.5.7")'
RUN Rscript -e 'remotes::install_version("shinydashboard",upgrade="never", version = "0.7.1")'
RUN Rscript -e 'remotes::install_version("RSQLite",upgrade="never", version = "2.2.3")'
RUN Rscript -e 'remotes::install_version("plotly",upgrade="never", version = "4.9.3")'
RUN Rscript -e 'remotes::install_version("leafgl",upgrade="never", version = "0.1.1")'
RUN Rscript -e 'remotes::install_version("lubridate",upgrade="never", version = "1.7.10")'
RUN Rscript -e 'remotes::install_version("leaflet.extras2",upgrade="never", version = "1.1.0")'
RUN Rscript -e 'remotes::install_version("golem",upgrade="never", version = "0.2.1")'
RUN Rscript -e 'remotes::install_version("DT",upgrade="never", version = "0.17")'
RUN Rscript -e 'remotes::install_version("dashboardthemes",upgrade="never", version = "1.1.3")'

EXPOSE 3838

CMD ["R", "-e", "shiny::runApp('/app', host = '0.0.0.0', port = 3838)"]`
